programs/x509/cert_write.c

```diff
@@ -803,6 +842,357 @@ int main( int argc, char *argv[] )
 
     return( ret );
 }
+
+/** Below magic pattern is used with ATCAECC508A demo application (or similar)
+ *  running on target to differentiate between user input and cert_write.exe.
+ */
+#define REMOTE_KEY_CMD_TAG          "remote"
+#define REMOTE_KEY_MAGIC_PATTERN    "rEmOtEkEy"
+#define REMOTE_KEY_ID_MIN           0
+#define REMOTE_KEY_ID_MAX           7
+#define REMOTE_KEY_SERIAL_BAUD      CBR_9600
+
+#define REMOTE_KEY_FUNC_GET_PUBKEY  0xA
+#define REMOTE_KEY_FUNC_SIGN        0xB
+
+extern mbedtls_pk_info_t mbedtls_eckey_info;
+
+int is_remote_key( const char * remote_info )
+{
+    size_t tag_len = strlen( REMOTE_KEY_CMD_TAG );
+    printf ("is_remote_key %s\n", remote_info);
+    if ( strlen( remote_info ) > tag_len &&
+            strncmp( remote_info, REMOTE_KEY_CMD_TAG, tag_len ) == 0 )
+        return 1;
+    return 0;
+}
+
+/** Load a transparent public key context with public key from remote device
+ *  over serial.
+ *  This function sends:
+ *      rEmOtEkEy<char encoded function code=GetPubKey><char encoded private key ID>
+ *  Receives:
+ *      <4 bytes length indicator in network order><concatenated public key>
+ */
+int load_pubkey_from_remote( const char * remote_info, mbedtls_pk_context * ctx )
+{
+    int key_idx = 0, offset = 0, ret = 0;
+    const char * serial_port = NULL;
+    unsigned char func_buffer[10];
+    unsigned char pub_key_buf[100];
+    size_t rx_len = 0;
+    static mbedtls_ecp_keypair ecp_key;
+
+    offset = strlen( REMOTE_KEY_CMD_TAG );
+    key_idx = (int)remote_info[offset++];
+    key_idx = key_idx - 48; // ascii to decimal
+
+    if ( key_idx < REMOTE_KEY_ID_MIN || key_idx > REMOTE_KEY_ID_MAX )
+    {
+        mbedtls_printf( " failed\n  !  Invalid remote key index %d\n\n", key_idx );
+        return( -1 );
+    }
+    serial_port = remote_info + offset;
+
+    /* Prepare command */
+    offset = 0;
+    func_buffer[offset++] = REMOTE_KEY_FUNC_GET_PUBKEY;
+    func_buffer[offset++] = key_idx;
+
+    if ( serial_xfer( serial_port, func_buffer, offset, pub_key_buf, sizeof( pub_key_buf ), &rx_len ) != 0 )
+    {
+        mbedtls_printf( " failed\n  !  Serial error trying to get pulic key\n\n" );
+        return( -1 );
+    }
+
+    /* Import public key from received binary */
+    mbedtls_ecp_keypair_init(&ecp_key);
+    ret = mbedtls_ecp_group_load(&ecp_key.grp, MBEDTLS_ECP_DP_SECP256R1);
+    if ( ret != 0 )
+        return( -1 );
+    ret = mbedtls_ecp_point_read_binary(&ecp_key.grp, &ecp_key.Q, pub_key_buf, rx_len );
+    if ( ret != 0 )
+    {
+        mbedtls_printf( " failed\n  !  Failed to read ecp key from binary\n\n" );
+        return( -1 );
+    }
+    ctx->pk_info = &mbedtls_eckey_info;
+    ctx->pk_ctx = &ecp_key;
+    return( 0 );
+}
+
+/**
+ * @brief           Tell if can do the operation given by type
+ *
+ * @param type      Target type
+ *
+ * @return          0 if context can't do the operations,
```

Comment:
Unlike other functions this returns a boolean result. 1 for can do, 0 for can't do. See builtin implementations in ```pk_wrap.c```.

Reply:
I wonder if the `can_do` method should return success/error instead of a boolean, It might need to communicate with the device and therefore want to return `HW_FAILED`. But let's not change it this week.